cips/cip-network-upgrade.md

```diff
@@ -0,0 +1,75 @@
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The next upgrade will be coordinated using a hard coded height that is passed as a flag when the node commences. This is an exception necessary to introduce the upgrading protocol which will come into affect for the following upgrade.
+
+The network introduces two new message types `MsgSignalVersion` and `MsgTryUpgrade`
+
+```protobuf
+message MsgSignalVersion {
+  string validator_address = 1;
+  uint64 version = 2;
+}
+
+message MsgTryUpgrade { string signer = 1; }
+```
+
+Only validators can submit `MsgSignalVersion` . The Celestia state machine tracks which version each validator has signalled for. The signalled version MUST either be the same version or the next. There is no support for skipping versions or downgrading.
+
+Clients may query the tally for each version as follows:
+
+```protobuf
+message QueryVersionTallyRequest { uint64 version = 1; }
+
+message QueryVersionTallyResponse {
+  uint64 voting_power = 1;
+  uint64 threshold_power = 2;
+  uint64 total_voting_power = 3;
+}
+```
+
+When `voting_power` is greater or equal to `theshold_power` , the network MAY upgrade. This is done through a “crank” transaction, `MsgTryUpgrade`, which can be submitted by any account. The account covers the gas required for the calculation. If the quorum is met, the chain will perform the necessary migrations at the end of processing that block. The proposer of the following height will include the new version in the block. As is currently, nodes will only vote for blocks that match the same network version as theirs.
+
+If the network agrees to move to a version that is not supported by the node, the node will gracefully shutdown.
+
+## Rationale
+
+When attempting a major upgrade, there is increased risk that the network halts. At least 2/3 in voting power of the network is needed to migrate and agree upon the next block. However this does not take into account the actions of byzantine validators. As an example, say there is 100 in total voting power. 66 have signalled v2 and 33 have yet to signal i.e. they are still signalling v1. It takes 1 byzantine voting power to pass the threshold, convincing the network to propose v2 blocks and then omitting their vote leaving the network failing to reach consensus until one of the 33 are able to upgrade. At the other end of the spectrum, increasing the necessary threshold means less voting power required to veto the upgrade. The middle point here is setting a quorum of 5/6ths which provides 1/6 byzantine fault tolerance to liveness and requiring at least 1/6th of the network to veto the upgrade.
```

Comment:
this is really good rationale for why the percentage is over 2/3s, but doesn't contian the rest of the rational for upgrades that don't use token voting or single binaries. I'm not blocking on it, but we might want to include more here from previous documents

Reply:
I would recommend discussions on the proposal to happen in a forum post that @cmwaters should provide. PRs get merged so convos get lost over time.

Reply:
SGTM and makes sense